Kill all the children, not just the firstborn

khoomeister opened the pull request:
process... that is.

Fixes #159

EndangeredMassa commented on src/processes/server.coffee:
```diff
@@ -168,9 +171,10 @@ spawnServer = (logs, name, cmd, args, opts, cb) ->
-        console.error err.stack
+      for childProcess in children
+        try childProcess.kill()
+        catch err
```
glad that `err` is different from `error`, but can we call this `childError` (or something like that) instead to be a little more clear?

khoomeister replied:
Sounds good - fixed.

johan commented:
:heart: :+1: :joy:

khoomeister commented:
Hey @EndangeredMassa, I've renamed the child error from `err` to `childErr`.

EndangeredMassa commented on src/processes/server.coffee:
```diff
@@ -168,9 +171,10 @@ spawnServer = (logs, name, cmd, args, opts, cb) ->
+        try childProcess.kill()
+        catch childErr
+          console.error childErr.stack
       throw error
```
This will add a new `process.on 'uncaughtException'` handler for each child. We probably only need one.

Also, we should add something to each `child.on 'exit'` that calls this `killAllChildren` logic as well.

khoomeister replied:
Wouldn't `process.on 'uncaughtException'` just add a handler on the testium runner process?  We're trying to kill phantom, proxy & webapp at the same time before exiting gracefully.

EndangeredMassa replied:
It does, but if a child crashes, it doesn't kill the parent process because we already add an error handler on the child. That handler does not kill other children or the parent process, right now.

khoomeister replied:
Ahh OK, sounds good - will update accordingly.

khoomeister replied:
OK, it's updated.  What do you think?

khoomeister commented:
I updated the code to include a "limit to one instance" check so the `process` object won't be filled up with more of the same listeners.

I also updated coffeescript to 1.9.3 to avoid this error when running `coffee -wcbo lib src`:

```
Error: The two following source files have the same output file:
    .../testium/src/processes/server.coffee
    .../testium/src/processes/server.coffee
```
Source: https://github.com/jashkenas/coffeescript/issues/3863

khoomeister commented:
OK tests added - actually discovered that what what we wanted was the `exit` event in the child process, not `error`.

So if any child process exits with a non-zero code, it'll kill everything.

Exit event doco: https://nodejs.org/api/child_process.html#child_process_event_exit

I read through the [error event doco](https://nodejs.org/api/child_process.html#child_process_event_error), but I'm not exactly sure how it can be triggered.  Perhaps you'd know more, @EndangeredMassa?

But aside from that, what do you think of the approach?

jkrems commented:
One example is [`ENOENT`](https://github.com/groupon/testium/blob/ck-kill-all-the-children/src/processes/server.coffee#L179). So passing a non-existent file as the command might do the trick.

khoomeister commented:
OK, `ENOENT` will only occur only if the `node`, `coffee` or `phantomjs` executables aren't available in the PATH... so I killed the PATH env variable :see_no_evil: in the test.

OK, how's it looking now?

EndangeredMassa commented on test/cleanup.test.coffee:
```diff
@@ -0,0 +1,76 @@
+
+getNumProcesses = (done) ->
+  exec 'ps', (err, stdout, stderr) ->
+    numProcesses = stdout.split('\n').length - 1 # header line
```
Can we break on error here?

```coffeescript
return done(err) if err?
```

EndangeredMassa commented on test/cleanup.test.coffee:
```diff
@@ -0,0 +1,76 @@
+  testium_logDirectory: LOG_DIRECTORY
+
+getNumProcesses = (done) ->
+  exec 'ps', (err, stdout, stderr) ->
```
I was afraid that this might make the test too flakey, but `ps` with no options actually sounds fine, I think.

> By default, ps selects all processes with the same effective user ID (euid=EUID) as the current user and associated with the same terminal as the invoker.

EndangeredMassa commented on test/cleanup.test.coffee:
```diff
@@ -0,0 +1,76 @@
+
+    mocha = execFile './node_modules/.bin/mocha', [ file ], {
+      env: extend(envOverrides, ENV_OVERRIDES, process.env)
+    }, (err, stdout, stderr) ->
```
can we break on this `err` as well?

EndangeredMassa commented on test/cleanup.test.coffee:
```diff
@@ -0,0 +1,76 @@
+{extend} = require 'lodash'
+
+LOG_DIRECTORY = "#{__dirname}/cleanup_log"
+ENV_OVERRIDES = 
```
Please remove extra whitespace at the end of this line.

EndangeredMassa commented on test/cleanup_dead_child_app.coffee:
```diff
@@ -0,0 +1,10 @@
+
+setTimeout ->
+  throw 'error!'
+, 2000
```
Do we need to wait this long? This time adds directly to the suite run time. This test has the same result if we kill it immediately.

EndangeredMassa commented on test/cleanup_exception_test.test.coffee:
```diff
@@ -0,0 +1,13 @@
+    setTimeout ->
+      assert.equal 0, 1
+      done()
+    , 1000
```
Do we need to wait this long?

khoomeister commented:
Hey @EndangeredMassa, thanks for the feedback - I've made all the changes :smiley:

EndangeredMassa commented:
published in [3.1.0](https://github.com/groupon/testium/releases/tag/v3.1.0)
